bpf/lib/ipip_termination.h

```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+
+
+#ifndef __LIB_IPIP_TERMINATION_H_
+#define __LIB_IPIP_TERMINATION_H_
+
+#ifdef ENABLE_IPIP_TERMINATION
+
+static __always_inline int
+decap_ipv4(struct __ctx_buff *ctx)
+{
+	struct iphdr *outer_ip4;
+	struct iphdr inner_ip4;
+	void *data, *data_end;
+
+	struct ipv4_ct_tuple tuple = {};
+	int ret, inner_ip4_off, outer_ip4_off = ETH_HLEN, outer_ip4_len, l4_off;
+
+	struct lb4_key key = {};
+	struct lb4_service *svc;
+
+	if (!revalidate_data_pull(ctx, &data, &data_end, &outer_ip4))
+		return DROP_INVALID;
+	if (outer_ip4->protocol != IPPROTO_IPIP) {
+		return CTX_ACT_OK;
+	}
+
+	outer_ip4_len = ipv4_hdrlen(outer_ip4);
+	inner_ip4_off = outer_ip4_off + outer_ip4_len;
+
+	if (ctx_load_bytes(ctx, inner_ip4_off, &inner_ip4, sizeof(inner_ip4)) < 0)
+		return DROP_INVALID;
+
+	ret = lb4_extract_tuple(ctx, &inner_ip4, inner_ip4_off, &l4_off, &tuple);
+	if (IS_ERR(ret)) {
+		/* Bypass */
+		return CTX_ACT_OK;
+	}
+
+	lb4_fill_key(&key, &tuple);
+
+	svc = lb4_lookup_service(&key, false);
+	if (svc) {
+		if (!lb4_svc_is_loadbalancer(svc))
+			return CTX_ACT_OK;
+
+		if (ctx_adjust_hroom(ctx, -outer_ip4_len, BPF_ADJ_ROOM_MAC, BPF_F_ADJ_ROOM_FIXED_GSO))
+			return DROP_INVALID;
+	}
+
+	return CTX_ACT_OK;
+}
+
+static __always_inline int
+decap_ipip(struct __ctx_buff *ctx)
```

Comment:
I now understand @borkmann's method of using tunnel netdev. While Daniel's approach is good, this PR does not require additional tunnel netdev and bpf_overlay attachment. The structure is simpler. So It makes debugging easier and providing performance advantages. It would be nice to have this feature as an option other than `--enable-ipip-termination`. Please discuss this with Daniel and let me know. I will be fixing some of the thing you mentioned above :)

Reply:
> The structure is simpler.

@gentoo-root I agree that the generic existing parts are more robust. However, my concerns are as follows.
- Changes for attaching `bpf_overlay` could be more complex, and could make the structure more complicated.
- This is because handling service IPs requires two filters: `bpf_host` and `bpf_overlay`. IPIP needs to be processed in `bpf_overlay`, and others should be handled in `bpf_host`.

> providing performance advantages

I expect that it will take a bit more time for the skb to reach the tunnel netdev via eth0. While the CPU instruction count for the decapsulation logic in the tunnel netdev and this PR might be similar, I believe more CPU instruction s will be consumed before the tunnel netdev decapsulation code is executed. Although the performance difference may not be significant, users may notice a slight difference when PPS(Packets Per Second) is high and packet sizes are small.


How about this approach?
This PR has been used in our private cloud on a 7K cluster for over a year, and it has also been presented at [KubeCon Europe 2024](https://events.linuxfoundation.org/kubecon-cloudnativecon-europe/program/schedule/), so I believe the code has been reasonably well-validated. And the IPv6 development is nearly complete.

How about we first merge the approach using `bpf_host` and then improve it by modifying `bpf_overlay` to attach the tunnel netdev in the future?

Reply:
@borkmann @gentoo-root Ping? 😄 Would you please review this PR?